### PR TITLE
Web Performance Optimization with webpack

### PR DESCRIPTION
**Pull request by iamakulov**
What's changed, or what was fixed?
- Added an article on front-end performance with webpack

**Target Live Date:** Probably January?

- [x] This has been reviewed and approved by @addyosmani 
- [x] I have run `gulp test` locally and all tests pass. (There’re a few `<script>`-related warnings, but these `<script>` blocks are code examples)
- [x] I have added the appropriate `type-something` label. (Not sure what’s this and haven’t found related info)
- [x] I've staged the site and manually verified that my content displays correctly.

**R:** @petele 

-----

I also have a few formatting-related questions:
1. For some reason, “Note:” blocks were recognized during one server launch and weren’t recognized during another. Probably that’s my local buggy installation, but IDK. Could we check this on the proper environment?

2. Locally, `{: .hide-from-toc }` wasn’t hiding “Further reading” headers from TOC (see e.g. [`decrease-frontend-size.md`](https://github.com/iamakulov/WebFundamentals/blob/webpack-article/src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md). Could you please check if I’m using them properly?

3. Is there a way to enable diff-like formatting (see below) for specific code blocks? I use it in a few code examples.

```diff
- const a = 5;
+ const b = 5;
```

4. I’d like to add “Next section” arrows like in [“Optimizing Content Efficiency”](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/). How to I do this? I haven’t found any related markup, and the “Optimizing Content Efficiency” article in my local build didn’t have them.
 
image

### CODE REVIEW

**Comment by iamakulov** [comment]
I signed it!

**Comment by iamakulov** [comment]
Also, a licensing question. In [`monitor-and-analyze.md`](https://github.com/iamakulov/WebFundamentals/blob/webpack-article/src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md), I’m using a screen recording from the [webpack-bundle-analyzer](https://github.com/webpack-contrib/webpack-bundle-analyzer) repo. I’ve converted it from GIF to MP4 (i.e. it’s probably a modification). The repo is licensed under MIT.

I’ve attributed the video source in a `<figcaption>`. Would this be enough to comply with the license? (From my understanding, this should be, but just in case Google has stricter requirements.)

**Comment by petele** [comment]
> For some reason, “Note:” blocks were recognized

`Note:` formatting looks for a specific regex, if it's not matched, it may not work. Our Dev server isn't perfect, but it should get it 95% of the time.

> Locally, {: .hide-from-toc } wasn’t hiding

Correct, it's only supported on production, not in our dev server

> Is there a way to enable diff-like formatting

No, sadly not. 

> I’d like to add “Next section” arrows

To add them, you need to add a `step_group: unique_id` to the `_toc.yaml` file for that section. It will show up in production, but isn't supported in the dev server.

Short version - our dev server simulates the real production environment, while it supports the majority of things that the production env supports, it doesn't support everything and it's always perfect.

**Comment by iamakulov** [comment]
> > Is there a way to enable diff-like formatting
> No, sadly not.

Replaced with “Before”–“After” comments.

> > I’d like to add “Next section” arrows
> To add them, you need to add a `step_group: unique_id` to the `_toc.yaml` file for that section.

Added!

File: src/content/en/fundamentals/performance/webpack/conclusion.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  Take a fresh look at your app’s performance at whole every few months
+
+Webpack is not the only tool that could help you make an app faster. Consider making your
+application [a Progressive Web App](/web/progressive-web-apps/) for even better experience and use
+automated profiling tools like [Lighthouse](/web/tools/lighthouse/) to get improvement suggestions.
+
+Don’t forget to read [webpack docs](https://webpack.js.org/guides/) – they have plenty of other
+useful information.
+
+And make sure to play with the training app **TODO: link to apply these principles in practice!**
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/conclusion.md]: is there supposed to be a link to the TODO app here?

File: src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+</pre>
+
+</li>
+</ol>
+
+Note: If you prefer to configure environment variables via CLI, take a look at the
+[EnvironmentPlugin](https://webpack.js.org/plugins/environment-plugin/). It works like the
+`DefinePlugin`, but reads the environment and replaces `process.env.` expressions automatically.
+
+Note: You’re not required to use the `UglifyJsPlugin`. You can use any different minifier as soon as
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md]: Three consecutive `notes` looks a little icky. Do they all need to be notes?

File: src/content/en/fundamentals/performance/webpack/index.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+webpack is one of the most popular bundling tools in use today. Taking advantage of its features for
+optimizing modern code, code-splitting scripts into critical and non-critical pieces and stripping
+out unused code (to name but a few optimizations) can ensure your app has a minimal network and
+processing cost.
+
+Let’s get started by looking at optimizing one of the costliest resources in a modern app –
+JavaScript.
+
+Note: We created a training app to play with optimizations described in this article. Try squeezing
+the most out of it to practice the tips **TODO: link**
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/index.md]: TODO link?

File: src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    };
+
+
+and run the production build. The plugin will open the stats page in a browser.
+
+Note: By default, the stats page shows the size of parsed files (i.e., of files as they appear in
+the bundle). You’ll likely want to compare gzip sizes since that’s closer to what real users
+experience. Use the sidebar on the left to switch the sizes.
+
+Note: If you use the
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md]: same comment about consecutive notes

File: src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+[ModuleConcatenationPlugin](https://webpack.js.org/plugins/module-concatenation-plugin/), it might
+merge a part of modules in the webpack-bundle-analyzer output, making the report less detailed. If
+you use this plugin, disable it during the analysis.
+
+Here’s what to look for in the report:
+
+- **Large dependencies.** Why are they so large? Are there smaller alternatives (e.g., Preact
+instead of React)? Do you use all the code it includes (e.g., Moment.js includes a lot of locales
+that are often not used and could be dropped TODO: link to the repo with Moment.js optimizations)?
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md]: todo link?

**Comment by petele** [comment]
@addyosmani - can you give an LGTM on this when it's ready to go please?

**Comment by addyosmani** [comment]
Sure thing. Initial PR LGTM and I’ll take another read through once the other comments have been addressed.

On the comment about consecutive notes, I feel the content there is important - perhaps we make them less unsightly by converting one or two to regular text rather than a highlighted note? :)

**Comment by iamakulov** [comment]
@petele @addyosmani Thanks for reviewing! :–)

I got rid of the consecutive notes. I had to remove the note about the webpack’s `-p` flag, but I didn’t really feel it fits well anyway. (Feel free to say if you still want to see in the article.)

The TODO links are not ready yet ­­­­– I’ll ping as soon as I add them.

**Comment by petele** [comment]
@iamakulov - since todo links aren't ready yet, i assume i should wait to merge and push live until they're ready?

File: src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+* [The UglifyJsPlugin docs](https://github.com/webpack-contrib/uglifyjs-webpack-plugin)
+
+* Other popular minifiers: [Babel
+  Minify](https://github.com/webpack-contrib/babel-minify-webpack-plugin), [Google Closure
+  Compiler](https://github.com/roman01la/webpack-closure-compiler)
+
+## Specify `NODE_ENV=production`
+
+Another way to decrease the front-end size is to set the `NODE_ENV`
+[environmental variable](https://superuser.com/questions/284342/what-are-path-
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md]: don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars

File: src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+</ol>
+
+Note: You’re not required to use the `UglifyJsPlugin`. You can use any different minifier as soon as
+it supports dead code removal (e.g., the [Babel Minify
+plugin](https://github.com/webpack-contrib/babel-minify-webpack-plugin) or the [Google Closure
+Compiler plugin](https://github.com/roman01la/webpack-closure-compiler)).
+
+### Further reading {: .hide-from-toc }
+
+* [What “environment variables” are](https://superuser.com/questions/284342/
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md]: don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars

File: src/content/en/fundamentals/performance/webpack/use-long-term-caching.md [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+    <script src="./vendor.26886caf15818fa82dfa.js"></script>
+    <script src="./main.00bab6fd3100008a42b0.js"></script>
+
+### Further reading {: .hide-from-toc }
+
+* Webpack guide [on long term caching](https://webpack.js.org/guides/caching/)
+
+* Webpack docs [about webpack runtime and manifest](https://webpack.js.org/concepts/manifest/)
+
+* [“Getting the most out of the CommonsChunkPlugin”](https://medium.com/webpack/
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/use-long-term-caching.md]: don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars

File: src/content/en/fundamentals/performance/webpack/use-long-term-caching.md [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+2, but for 30 chunks, it might be 8 – because if you keep it at 2, too many modules will get into
+the common file, inflating it too much.
+
+### Further reading {: .hide-from-toc }
+
+* Webpack docs [about the concept of entry points](https://webpack.js.org/concepts/entry-points/)
+
+* Webpack docs [about the CommonsChunkPlugin](https://webpack.js.org/plugins/commons-chunk-plugin/)
+
+* [“Getting the most out of the CommonsChunkPlugin”](https://medium.com/webpack/webpack-bits-
```
Comment (petele) [inline src/content/en/fundamentals/performance/webpack/use-long-term-caching.md]: don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars. 

if you really wanted to line break this, you could do something like:

```
* [“Getting the most out of the
CommonsChunkPlugin”](https://medium.com/webpack/webpack-bits-more-and-more-and-more-and-more)
```

Comment (iamakulov) [reply src/content/en/fundamentals/performance/webpack/use-long-term-caching.md]: > don't linebreak URLs, otherwise the link breaks

OK! (I checked locally, and they worked, but I’ll fix them.) Will fix this shortly.

> if you really wanted to line break this, you could do something like:

I’ve actually got test fails because of this: https://travis-ci.org/google/WebFundamentals/builds/325794409#L743. (If you see [the line which triggers the error](https://github.com/google/WebFundamentals/commit/3c3ac29abc1ce4380e9fcafabf9e325dcd85ed8f#diff-9f55b5ed9c2b6c50e93df9647ab49499R158), there’s a line break inside the link name.)

**Comment by iamakulov** [comment]
> since todo links aren't ready yet, i assume i should wait to merge and push live until they're ready?

Yes, exactly! Thanks.

**Comment by addyosmani** [comment]
The fixes for consecutive notes lgtm. It looks like we're still working on the line-break issues and TODOs here. Will take another read through once @iamakulov is happy with those have been addressed :)

**Comment by petele** [comment]
@addyosmani - i've added the `DO NOT MERGE` tag, i'll ignore this until you remove it and let me know it's good to go!

**Comment by addyosmani** [comment]
@iamakulov would it be possible to get an ETA on when the above remaining edits can be completed? :)

**Comment by iamakulov** [comment]
@addyosmani I hate these delays, sorry so much for them :–(

Will fix line breaks on this Saturday. ETA on TODOs is the end of the next week (the preview should be earlier)

**Comment by addyosmani** [comment]
@iamakulov Thanks! Took a look through the latest commit. I'll keep an eye out for any further updates to the TODOs.

**Comment by iamakulov** [comment]
(JFYI: made a couple of minor changes)

**Comment by addyosmani** [comment]
Awesome. Thanks for the updates!

**Comment by iamakulov** [comment]
And... we’re done here!

(The links in the latest commit are offline right now, but they will become live as soon as we move the repos.)

**Comment by addyosmani** [comment]
The repos for this project have been approved for launch. We're going to get them moved over to the GoogleChromeLabs org today and should be good to land this content at any time before/between. This should be fine as long as we're not tweeting the content out yet (I'd aim to get that done tomorrow).

**Comment by iamakulov** [comment]
Added article links to https://github.com/GoogleChromeLabs/webpack-training-project.

I think we’re done.
